src/librustc/traits/select.rs

```diff
@@ -2425,7 +2426,42 @@ impl<'cx, 'gcx, 'tcx> SelectionContext<'cx, 'gcx, 'tcx> {
                                        expected_trait_ref.clone(),
                                        obligation_trait_ref.clone())
             .map(|InferOk { obligations, .. }| self.inferred_obligations.extend(obligations))
-            .map_err(|e| OutputTypeParameterMismatch(expected_trait_ref, obligation_trait_ref, e))
+            .map_err(|e| {
+                let self_ty = expected_trait_ref.self_ty();
+                match (&self_ty.sty, &e) {
```

Comment:
Can you please just compare the # of parameters in the fn sig vs. the # of parameters in the obligation (that's `obligation.predicate.skip_binder().subst_at(1)`)?